extensions/amp-story-shopping/0.1/amp-story-shopping-config.js

```diff
@@ -14,6 +17,108 @@ import {
  */
 let ShoppingConfigResponseDef;
 
+/**
+ * Validates string length of shopping config attributes
+ * @param {?string} field
+ * @param {?string} value
+ * @return {!Array<string>}
+ * @private
+ */
+function validateRequired(field, value) {
+  if (value === undefined) {
+    return [`Field ${field} is required.`];
+  }
+}
+
+/**
+ * Validates string length of shopping config attributes
+ * @param {?string} field
+ * @param {?string} str
+ * @param {?number} maxLen
+ * @return {!Array<string>}
+ * @private
+ */
+function validateStringLength(field, str, maxLen = 100) {
+  if (str !== undefined && str.length > maxLen) {
+    return [`Length of ${field} exceeds max length: ${str.length} > ${maxLen}`];
+  }
+}
+
+/**
+ * Validates number in shopping config attributes
+ * @param {?string} field
+ * @param {?number} number
+ * @private
+ * @return {!Array<string>}
+ */
+function validateNumber(field, number) {
+  if (number !== undefined && isNaN(number)) {
+    return [`Value for field ${field} is not a number`];
+  }
+}
+
+/**
+ * Validates url of shopping config attributes
+ * @param {?string} field
+ * @param {!Array<string>} url
+ * @private
+ * @return {!Array<string>}
+ */
+function validateURL(field, url) {
+  if (url === undefined) {
+    /* Not a Required Attribute, return.*/
+    return;
+  }
+
+  const urls = Array.isArray(url) ? url : [url];
+
+  return urls.map((url) => {
+    try {
+      assertHttpsUrl(url, `amp-story-shopping-config ${field}`);
+    } catch (e) {
+      return e.message;
+    }
+  });
+}
+
+/** @private @const {!Object<Array<Function>>} */
+const productValidationConfig_ = {
+  /* Required Attrs */
+  'productId': [validateRequired, validateStringLength],
+  'productTitle': [validateRequired, validateStringLength],
+  'productPrice': [validateRequired, validateNumber],
+  'productImages': [validateRequired, validateURL],
+  'productPriceCurrency': [validateRequired, validateStringLength],
+  /* Optional Attrs */
+  'productColor': [validateStringLength],
+  'productSize': [validateStringLength],
+  'productIcon': [validateURL],
+  'productTagText': [validateStringLength],
+  'reviewsData': [validateURL],
+  'ctaText': [validateNumber],
+  'shippingText': [validateNumber],
+};
+
+/**
+ * Validates shopping config.
+ * @param {!ShoppingConfigDataDef} shoppingConfig
+ * @private
+ */
+function validateConfig_(shoppingConfig) {
```

Comment:
This function's logic is a bit tricky to parse. An alternative approach could be to set up the other validation functions to do nothing except throw an error in the failure case. For example:
```
function validateRequired(field, value) {
  if (value === undefined) {
    throw Error(`Field ${field} is required.`);
  }
}
```

Once the other validation functions throw errors, this function's logic becomes simpler because you can then loop through all the function calls with a try/catch and call `user().warn()` whenever an error is caught.

```
Object.keys(productValidationConfig_).forEach((configKey) => {
  const validationFunctions = productValidationConfig_[configKey];
  validationFunctions.forEach((fn) => {
    try {
      fn(configKey, shoppingConfig[configKey]);
    } catch (err) {
      user().warn('amp-story-shopping-config', err);
    }
  });
});
```
Thoughts about taking this approach?

Reply:
This approach seems succinct and easy-to-read - I've made the change!